feat: Form Builder

shariquerik opened the pull request:
This feature aims to help the user edit the DocType's fields layout directly from UI (Form Builder), rather than from the form view of the DocType.

Triggered by https://github.com/frappe/frappe/pull/17785

**Key Features:**
- [x] Drag & Drop fields from the sidebar to form
- [x] Add New Tabs, Sections & Columns
- [x] Delete Tabs, Sections, Columns & Fields
- [x] Edit the details of Tabs, Sections, Columns & Fields
- [x] Drag & Drop to move around sections & fields in the form and across tabs
- [x] Save and update the changes
- [x] Should also work for Customize Form fields to add and update custom fields
- [x] Double click to edit the label
- [x] Move current field/column/section and the following to a new column/section/tab
- [x] Render different field types (e.g Data, Select, Check etc.) in the properties section as well as on the form.
- [x] Preview Mode
- [x] Make options field a Link Field for Table & Link type fields
- [x] Fetch From (dynamically get doctype list and fieldname list in select field)
- [x] Validation Before Save
- [x] UI Test

Check it out on YouTube

[![IMAGE_ALT](https://img.youtube.com/vi/hUBB5MM4O8k/maxresdefault.jpg)](https://www.youtube.com/watch?v=hUBB5MM4O8k)


https://user-images.githubusercontent.com/30859809/213107164-e19b9313-a12c-484b-a303-e50d04ad4879.mov


>no-docs

shariquerik committed:
feat: Form Builder

- Form Builder Page
- Vue 3 setup
- Pinia store setup
- Reusable Tab Component
- Form Builder button on Doctype

shariquerik committed:
revert: form editor code

shariquerik committed:
fix: added all_fieldtype list in frappe.model

Alchez commented on frappe/public/js/form_builder/store.js:
```diff
@@ -0,0 +1,38 @@
+		is_dirty() {
+			return this.dirty;
+		},
+		fetch() {
```
This could just become an `async` function, right? That may help cleanup the callbacks on `frappe.model.with_doctype` and `nextTick`

shariquerik committed:
chore: code formatting

shariquerik committed:
fix: removed Reusable Tab Component

Not very reusable. Setup tabs for sidebar

shariquerik committed:
fix: setup tabs, sections, columns & fields layout

shariquerik committed:
fix: hover and selected logic

shariquerik committed:
fix: remove field logic

shariquerik committed:
style: make tab content scrollable, sidebar & minor css changes

shariquerik committed:
fix: render fieldtypes in sidebar and make it draggable into sections

shariquerik committed:
fix: using seperate component for search box

shariquerik committed:
fix: show fields properties in field properties section in sidebar

shariquerik committed:
style: scrolling & sticky header for layout & sidebar

shariquerik committed:
style: applied field css to ghost element

shariquerik committed:
fix: on removing field switch to fieldtypes section

shariquerik committed:
style: minor fix

shariquerik committed:
fix: make page dirty if layout is updated

shariquerik committed:
fix: get computed layout -> if we reset changes layout gets updated

shariquerik committed:
fix: clear selected_field if field is removed

shariquerik committed:
chore: better naming

shariquerik committed:
fix: reset scroll bar css to default for form-builder page

shariquerik committed:
fix: scroll if field or section is dragged to top or bottom

shariquerik committed:
feat: resizable sidebar

shariquerik committed:
refactor: sidebar & tab layout

shariquerik committed:
fix: drag field & sections across tabs

shariquerik committed:
feat: add/remove column and section

shariquerik committed:
feat: add/remove tabs

shariquerik committed:
fix: fetch latest changes when we switch back to form builder page

shariquerik committed:
refactor: get new docfield's default fields for all newly added fields

shariquerik committed:
fix: confirm dialog before removing tab & section

shariquerik committed:
feat: save and update the changes

shariquerik committed:
fix: no need of new_field flag

shariquerik committed:
fix(minor): add section above

shariquerik committed:
fix: show alert while saving if page is not dirty

shariquerik committed:
feat: Update Customize Form Fields

- Handle route for both Doctype Fields & Customize Form Fields
- Minor changes in Doctype Field fetch logic
- Fetch & Save Customize Form Fields

shariquerik committed:
fix: added switch buttons on customize form & form builder

shariquerik committed:
fix: skip first tab, section & column if its new without data while saving

shariquerik committed:
fix: highlight custom fields and do not allow non custom fields to sort

shariquerik committed:
refactor: skip first tab/section/column logic

shariquerik committed:
fix: validation while removing standard fields

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
fix: use actual doctype name for routing instead of slugged version

shariquerik committed:
feat: make page read only if editing is restricted

shariquerik committed:
fix: show read only indicator

git-avc commented:
Amazing work @shariquerik !
I don't know if this PR is the right place to note issues ... If don't, please, let me know where I should do it.

In "Options" textbox CR is not allowed, so you can't specify the options to show in a "select" field.

![image](https://user-images.githubusercontent.com/94137451/197401123-4e2db764-e45c-47b6-b418-6664ec413afb.png)

git-avc commented:
Suggestions: 

- It would be great if you can drag fields between tabs, even the destination tab has not fields inside. Now, if you save the form, and a tab is empty, this tab can't "receive" any field.

- It would be great if you can drag the data fields from a list on the sidebar, that contains all the fields of the doctype that are not assigned yet to the form.

shariquerik commented:
> In "Options" textbox CR is not allowed, so you can't specify the options to show in a "select" field.

Thanks for pointing this out but I have plans to use different field types in the properties section and in the main area.

shariquerik commented:
> It would be great if you can drag fields between tabs, even the destination tab has not fields inside. Now, if you save the form, and a tab is empty, this tab can't "receive" any field.

I will add this in my list thanks

> It would be great if you can drag the data fields from a list on the sidebar, that contains all the fields of the doctype that are not assigned yet to the form.

I didn't understand this one

git-avc commented:
> > It would be great if you can drag the data fields from a list on the sidebar, that contains all the fields of the doctype that are not assigned yet to the form. 
> 
> I didn't understand this one

Imagine this. List current doctype fields on sidebar, and let drag and drop from the sidebar to the form ... Maybe useful if you define the field data (type, restrictions, etc ..) by traditional way, but use form builder just to place it on the form .... 

![image](https://user-images.githubusercontent.com/94137451/198029183-4a4b1661-88a0-4061-9193-db9e0fe349c9.png)


Thanks again. Form builder is a game changer feature.

shariquerik committed:
fix: use v-html for svg icons

shariquerik committed:
fix: use store in utils

shariquerik committed:
feat: add/remove/edit & sort columns

before this we columns were automatically created and there was no way to edit the name or sort columns

shariquerik committed:
refactor: keep related actions and code in same vue file

avoided using emit

shariquerik committed:
fix: using buttons for section actions instead of dropdown

since we only have two actions and to make consistent with tab, field & columns action

shariquerik committed:
fix: lose field focus if click outside form

shariquerik committed:
fix: remove tab/section/column/field logic

shariquerik committed:
fix: no need of remove flag

shariquerik commented:
> Imagine this. List current doctype fields on sidebar, and let drag and drop from the sidebar to the form ... Maybe useful if you define the field data (type, restrictions, etc ..) by traditional way, but use form builder just to place it on the form

No need for the `Fields` section in the sidebar. This feature is used to add new fields, update/sort existing fields or delete them.
According to you, we will have to first go to doctype add multiple fields and then come back here and add those fields in the form from the fields section why? All the fields are already available in the form

Thanks for spending time testing and for your suggestions.😊

shariquerik committed:
fix: make tabs header scrollable if has more tabs

shariquerik committed:
fix: diasble sort for column if column is standard in customize form

shariquerik committed:
fix: do not move section & columns if empty while removing tab & section

shariquerik committed:
fix: drag & drop or add new section if tab is empty

shariquerik committed:
fix: create tab/section/column break with unique name

shariquerik committed:
fix: do not allow to move the custom tab/section/column if it has standard field inside

shariquerik committed:
refactor: removed is_custom getter & refactored code accordingly

shariquerik committed:
fix: hide column action in read only mode

shariquerik committed:
fix: cannot lose focus from field input if non draggable element is clicked

also section title input field was disabled for non draggable elemnts

shariquerik committed:
fix: disable section title & hide tab header if there is none if read_only

shariquerik committed:
fix: change tab-content height based on tab-header

shariquerik committed:
fix: using watchEffect instead of multiple watch

shariquerik committed:
fix: cannot select section in read_only state

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
fix: if error occurred save button gets disabled

shariquerik committed:
chore: separated code into functions

shariquerik committed:
fix: hide/make read_only property fields based on read_only expression

shariquerik committed:
fix: make column border as dashed and fieldtypes border solid

shariquerik committed:
fix: make section collapsible

shariquerik committed:
fix: edit tab/section label on double click

shariquerik committed:
fix: hide border if hide_border is set

shariquerik committed:
fix: remove empty sections while loading and saving

shariquerik committed:
fix: keep first column if its is empty

shariquerik committed:
fix: handle cur_frm in depends_on expression

shariquerik committed:
fix: do not render hidden property fields

shariquerik committed:
fix: hide customize form btn & show form builder btn for child table doctypes

shariquerik committed:
fix: created EditableInput for tab & section label

shariquerik committed:
fix: lose focus if clicked outside app

shariquerik committed:
chore: active border style refactor

shariquerik committed:
fix: do not remove first section/tab if edited

shariquerik committed:
feat: move current field/column/section and the following to a new column/section/tab

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
fix: if custom section has empty standard column move empty column to previous section

shariquerik committed:
fix: no need to set random fieldname while adding column

shariquerik committed:
fix: show red asterisk sign if field is reqd

shariquerik committed:
fix: use EditableInput for field input

shariquerik committed:
feat: separate component for each fieldtypes used in properties & main form

Worked on Data & Check Component. Components are created for each fieldtypes but each is still rendering Data Components template

shariquerik committed:
fix: added/updated Data,Check,Select & Small Text Control Components

shariquerik committed:
fix: render rest of the fieldtypes in properties section & main form

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
fix: remove label focus outline while editing

also added tooltip to know double click to edit label

shariquerik committed:
fix: drag animation & cannot drag alert on dragging standard fields

shariquerik committed:
feat: preview mode

shariquerik committed:
fix: tab containing standard fields was draggable in customize form

shariquerik committed:
fix: check if field has standard field not if field is custom

shariquerik committed:
feat: make options field as link field for Table & Link type fields

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
fix: reset link field filters

shariquerik committed:
fix: do not use cur_frm in depends_on expression

cur_frm is deprecated

shariquerik committed:
fix: use onClickOutside feat from @vueuse/core

shariquerik committed:
fix: hide fetch_from field for no_value_type fields

shariquerik committed:
chore: rearranged code

shariquerik committed:
fix: dynamically get fetch_form doctype & fieldnames as select field

shariquerik committed:
fix: use df.read_only to make field disabled

shariquerik committed:
revert: cannot drag alert on dragging standard fields

not working as expected

shariquerik committed:
fix: validate before saving

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
test: UI Test for basic form builder features

shariquerik committed:
test: using custom doctype to test form builder

shariquerik committed:
chore: removed unsed code

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
fix: updated add/remove & move icon

shariquerik committed:
fix: support dark mode

shariquerik committed:
fix: inheret bg color for editable input

shariquerik committed:
Merge branch 'develop' into form-builder-vue3

shariquerik committed:
fix: Change doctype by clicking on page title

also added go to doctype list menu item

shariquerik committed:
test(UI): added more UI tests

shariquerik committed:
test(UI): force clear text

shariquerik committed:
Merge branch 'develop' into form-builder-vue3
